lib/ret/hub.ex

```diff
@@ -298,41 +353,70 @@ defmodule Ret.Hub do
 end
 
 defimpl Canada.Can, for: Ret.Account do
+  alias Ret.{Hub}
+  @object_actions [:spawn_and_move_media, :spawn_camera, :spawn_drawing, :pin_objects]
+  @special_actions [:update_hub, :close_hub, :embed_hub, :kick_users, :mute_users] ++ @object_actions
+
   # Always deny access to non-enterable hubs
   def can?(%Ret.Account{}, :join_hub, %Ret.Hub{entry_mode: :deny}), do: false
 
+  # Bound hubs - Join perm
   def can?(%Ret.Account{} = account, :join_hub, %Ret.Hub{hub_bindings: hub_bindings})
       when hub_bindings |> length > 0 do
     hub_bindings |> Enum.any?(&(account |> Ret.HubBinding.member_of_channel?(&1)))
   end
 
+  # Bound hubs - Manage actions
   def can?(%Ret.Account{} = account, action, %Ret.Hub{hub_bindings: hub_bindings})
       when action in [:update_hub, :close_hub] and hub_bindings |> length > 0 do
     hub_bindings |> Enum.any?(&(account |> Ret.HubBinding.can_manage_channel?(&1)))
   end
 
+  # Bound hubs - Moderator actions
   def can?(%Ret.Account{} = account, action, %Ret.Hub{hub_bindings: hub_bindings})
       when hub_bindings |> length > 0 and action in [:kick_users, :mute_users] do
     hub_bindings |> Enum.any?(&(account |> Ret.HubBinding.can_moderate_users?(&1)))
   end
 
+  # Bound hubs - Object permissions
+  def can?(%Ret.Account{} = account, action, %Ret.Hub{hub_bindings: hub_bindings} = hub)
+      when hub_bindings |> length > 0 and action in @object_actions do
+    is_moderator = hub_bindings |> Enum.any?(&(account |> Ret.HubBinding.can_moderate_users?(&1)))
+
+    if is_moderator do
+      true
+    else
+      is_member = hub_bindings |> Enum.any?(&(account |> Ret.HubBinding.member_of_channel?(&1)))
+      is_member and hub |> Hub.has_perm!(action)
+    end
+  end
+
+  # Bound hubs - Always prevent embedding
   def can?(%Ret.Account{}, action, %Ret.Hub{hub_bindings: hub_bindings})
       when hub_bindings |> length > 0 and action in [:embed_hub],
       do: false
 
-  # Anyone can join an unbound hub
-  def can?(_, :join_hub, %Ret.Hub{hub_bindings: []}), do: true
+  # Unbound hubs - Anyone can join an unbound hub
+  def can?(_account, :join_hub, %Ret.Hub{hub_bindings: []}), do: true
 
-  # Creators of unbound hubs can perform special actions
+  # Unbound hubs - Creators can perform special actions
   def can?(%Ret.Account{account_id: account_id}, action, %Ret.Hub{created_by_account_id: account_id})
-      when account_id != nil and action in [:update_hub, :close_hub, :embed_hub, :kick_users, :mute_users],
+      when account_id != nil and action in @special_actions,
       do: true
 
+  # Unbound hubs - Object permissions for regular users are based on perms settings
+  def can?(_account, action, hub) when action in @object_actions do
```

Comment:
I think I mentioned this before, but I feel like this module is getting long enough we may want to explicitly check for no bindings in the guard clause. I'm nervous about the inadvertent fallthrough risk.